domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+	GenerateUserActivationKey(context.Context, string) error
+	// RemoveUserActivationKey removes any activation key ass the given user
+	RemoveUserActivationKey(context.Context, string) error
+}
+
+// Service provides the API for working with upgrade info
+type Service struct {
+	st State
+}
+
+const (
+	// usernameValidationRegex is the regex used to validate that user names are
+	// valid for consumption by Juju. Usernames must be 1 or more runes long,
+	// can contain any unicode rune from the letter class and may contain zero
+	// or more of .,+ or - runes as long as they are don't appear at the start
+	// or end of the username.
+	usernameValidationRegex = "^([\\pL\\pN]|[\\pL\\pN][\\pL\\pN.+-]{0,253}[\\pL\\pN])$"
+)
+
+// NewService returns a new Service for interacting with the underlying state.
+func NewService(st State) *Service {
+	return &Service{st: st}
+}
+
+// GetUser will find and return the user associated with name. If there is no
+// user for the user name then an error that satisfies usererrors.NotFound will
+// be returned. If supplied with a invalid user name then a error that satisfies
+// usererrors.
+//
+// GetUser will not return users that have been previously deleted or removed.
+func (s *Service) GetUser(
+	ctx context.Context,
+	name string,
+) (user.User, error) {
+	if err := ValidateUsername(name); err != nil {
+		return user.User{}, fmt.Errorf("username %q: %w", name, err)
+	}
+
+	u, err := s.st.GetUser(ctx, name)
+	if err != nil {
+		return user.User{}, fmt.Errorf("getting user %q: %w", name, err)
+	}
+
+	return u, nil
+}
+
+// ValidateUsername takes a user name and validates that the user name is
+// conformant to our regex rules defined in usernameValidationRegex. If a user
+// name is not valid a error is returned that satisfies
+// usererrors.UsernameNotValid.
+//
+// Usernames must be 1 or more runes long, can contain any unicode rune from the
+// letter class and may contain zero or more of .,+ or - runes as long as they
+// are don't appear at the start or end of the username.
+func ValidateUsername(name string) error {
+	regex, err := regexp.Compile(usernameValidationRegex)
+	if err != nil {
+		return fmt.Errorf("compiling user name validation regex %q: %w",
+			usernameValidationRegex, err,
+		)
+	}
+	if !regex.MatchString(name) {
+		return fmt.Errorf("%w %q", usererrors.UsernameNotValid, name)
+	}
+	return nil
+}
+
+// AddUser will add a new user to the database. If the user already exists an
+// error that satisfies usererrors.AlreadyExists will be returned.
+func (s *Service) AddUser(ctx context.Context, name, displayName, password, creator string) error {
+	err := ValidateUsername(name)
+	if err != nil {
+		return fmt.Errorf("username %q: %w", name, err)
+	}
+
+	err = s.st.AddUser(ctx, name, displayName, password, creator)
+	if err != nil {
+		return fmt.Errorf("adding user %q: %w", name, err)
+	}
+	return nil
+}
+
+// RemoveUser marks the user as deleted. This obviates the ability of a user
```

Comment:
@cderici "marks the user as removed"

I think we want to completely remove the the last sentence copied over from Mongo state here and write a better description of what this service func is doing. Probably take a look at some of the examples above:

Things that should be included:
- What are the common error case error types that are going to be removed?
- We should state that when removing a user we are going to delete all auth information for the user.